mailpoet/tasks/phpstan/phpstan.neon

```diff
@@ -44,7 +44,7 @@ parameters:
     - '/Call to method getName\(\) on an unknown class _generated\\([a-zA-Z])*Cookie/' # codeception generate incorrect return type in ../../tests/_support/_generated
     -
       message: "#^Cannot cast string|void to string\\.$#"
-      count: 10
```

Comment:
I am not aware of a good way. You could do something like 
```
$query = $this->wpdb->prepare(<query>);
if (!is_string($query)) {
  // some error handling
}
```

But this seems quite a overhead.

Reply:
Agreed, that would be very annoying to have to do every time.

I'm far from a PHPStan expert, but it seems like one option would be to register a custom return type extension. As far as I can tell, `prepare` will only return `void` if it receives a `null` value for `$query`, so something like this should work:

```php
<?php declare(strict_types = 1);

namespace MailPoet\PHPStan\Extensions\CodeceptionExtension\Type;

use PhpParser\Node\Expr\MethodCall;
use PHPStan\Analyser\Scope;
use PHPStan\Reflection\MethodReflection;
use PHPStan\Type\DynamicMethodReturnTypeExtension;
use PHPStan\Type\NullType;
use PHPStan\Type\StringType;
use PHPStan\Type\Type;
use PHPStan\Type\VoidType;

class WPDBPrepareMethodReturnTypeExtension implements DynamicMethodReturnTypeExtension {
  public function getClass(): string {
    return 'wpdb';
  }

  public function isMethodSupported(MethodReflection $methodReflection): bool {
    return $methodReflection->getName() === 'prepare';
  }

  public function getTypeFromMethodCall(MethodReflection $methodReflection, MethodCall $methodCall, Scope $scope): Type {
    if (count($methodCall->args) > 0) {
      $argumentType = $scope->getType($methodCall->args[0]->value);

      if ($argumentType instanceof NullType) {
        return new VoidType();
      }

      return new StringType();
    }

    return new VoidType();
  }
}
```

Then we register it in `extensions.neon`:
```
services:
  # resolves return type from Stub::make(SomeMockedClass::class) and other Stub methods
  -
    class: MailPoet\PHPStan\Extensions\CodeceptionExtension\Type\StubDynamicReturnTypeExtension
    tags:
      - phpstan.broker.dynamicStaticMethodReturnTypeExtension

  # resolves return type from $this->make(SomeMockedClass::class) and other test case methods
  -
    class: MailPoet\PHPStan\Extensions\CodeceptionExtension\Type\TestCaseDynamicReturnTypeExtension
    tags:
      - phpstan.broker.dynamicMethodReturnTypeExtension
  -
    class: MailPoet\PHPStan\Extensions\CodeceptionExtension\Type\WPDBPrepareMethodReturnTypeExtension
    tags:
        - phpstan.broker.dynamicMethodReturnTypeExtension
```

When I tried this out locally, I got the results I was hoping for:

```
 1557/1557 [▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓] 100%

 ------ ------------------------------------------------------------------------------- 
  Line   lib/Automation/Engine/Storage/AutomationRunLogStorage.php                      
 ------ ------------------------------------------------------------------------------- 
  49     Else branch is unreachable because ternary operator condition is always true.  
 ------ ------------------------------------------------------------------------------- 

 ------ ----------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Automation/Engine/Storage/AutomationRunStorage.php                                                                       
 ------ ----------------------------------------------------------------------------------------------------------------------------- 
         Ignored error pattern #^Cannot cast string|void to string\.$# in path                                                        
         /Users/johnoleksowicz/Workspace/mailpoet/mailpoet/lib/Automation/Engine/Storage/AutomationRunStorage.php was not matched in  
         reported errors.                                                                                                             
  207    Else branch is unreachable because ternary operator condition is always true.                                                
 ------ ----------------------------------------------------------------------------------------------------------------------------- 

 ------ -------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Automation/Engine/Storage/AutomationStorage.php                                                                       
 ------ -------------------------------------------------------------------------------------------------------------------------- 
         Ignored error pattern #^Cannot cast string|void to string\.$# in path                                                     
         /Users/johnoleksowicz/Workspace/mailpoet/mailpoet/lib/Automation/Engine/Storage/AutomationStorage.php was not matched in  
         reported errors.                                                                                                          
 ------ -------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Automation/Integrations/WooCommerce/Fields/CustomerOrderFieldsFactory.php                                                  
 ------ ------------------------------------------------------------------------------------------------------------------------------- 
         Ignored error pattern #^Cannot cast string|void to string\.$# in path                                                          
         /Users/johnoleksowicz/Workspace/mailpoet/mailpoet/lib/Automation/Integrations/WooCommerce/Fields/CustomerOrderFieldsFactory.p  
         hp was not matched in reported errors.                                                                                         
 ------ ------------------------------------------------------------------------------------------------------------------------------- 

 ------ ------------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Automation/Integrations/WooCommerce/Fields/CustomerReviewFieldsFactory.php                                                 
 ------ ------------------------------------------------------------------------------------------------------------------------------- 
         Ignored error pattern #^Cannot cast string|void to string\.$# in path                                                          
         /Users/johnoleksowicz/Workspace/mailpoet/mailpoet/lib/Automation/Integrations/WooCommerce/Fields/CustomerReviewFieldsFactory.  
         php was not matched in reported errors.                                                                                        
 ------ ------------------------------------------------------------------------------------------------------------------------------- 

 ------ ---------------------------------------------------------------------------------------------------------------------------- 
  Line   lib/Automation/Integrations/WooCommerce/Fields/TermParentsLoader.php                                                        
 ------ ---------------------------------------------------------------------------------------------------------------------------- 
         Ignored error pattern #^Cannot cast string|void to string\.$# in path                                                       
         /Users/johnoleksowicz/Workspace/mailpoet/mailpoet/lib/Automation/Integrations/WooCommerce/Fields/TermParentsLoader.php was  
         not matched in reported errors.                                                                                             
 ------ ---------------------------------------------------------------------------------------------------------------------------- 

                                                                                                                        
 [ERROR] Found 7 errors                                                                                                 
                                                                                                                        

 [Exec]  Exit code 1  Time 52.19s
 1/1 [============================] 100%
```

Just as a test I tried adding some code that explicitly called `$wpdb->prepare(null)`, and PHPStan caught it: `Result of method wpdb::prepare() (void) is used.`

I think this would also mean not having to perform the cast to string in the first place since PHPStan should see it as a string.

If we want to make this change it would probably be better to do it as a separate PR. I do think it would make our lives a tiny bit easier when working directly with `wpdb`. Curious to hear your thoughts.

Reply:
I love it! Thank you @johnolek. I prepared a ticket [MAILPOET-5612]